### ISSUE TITLE
Azure: Failed to start CoreOS Metadata Agent

### ISSUE THREAD

**Issue by dtretyakov** [coreos/afterburn#109]
`coreos-metadata.service` does not start on Core OS in Azure.

## Operating System Version ##
```
> cat /etc/os-release
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1800.5.0
VERSION_ID=1800.5.0
BUILD_ID=2018-07-28-2250
PRETTY_NAME="Container Linux by CoreOS 1800.5.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```

## Environment ##
Azure VM

## Expected Behavior ##
`coreos-metadata.service` starts

## Actual Behavior ##

`coreos-metadata.service` fails

## Other Information ##

```
> sudo journalctl -u coreos-metadata.service
-- Logs begin at Thu 2018-08-02 14:43:17 UTC, end at Thu 2018-08-02 15:03:11 UTC. --
Aug 02 14:47:20 teamcity systemd[1]: Starting CoreOS Metadata Agent...
Aug 02 14:47:20 teamcity coreos-metadata[1503]: Aug 02 14:47:20.742 INFO Fetching http://168.63.129.16/?comp=versions: Attempt #1
Aug 02 14:47:20 teamcity coreos-metadata[1503]: Aug 02 14:47:20.765 INFO Fetch successful
Aug 02 14:47:20 teamcity coreos-metadata[1503]: Aug 02 14:47:20.765 INFO Fetching http://168.63.129.16/machine/?comp=goalstate: Attempt #1
Aug 02 14:47:20 teamcity coreos-metadata[1503]: Aug 02 14:47:20.769 INFO Fetch successful
Aug 02 14:47:20 teamcity coreos-metadata[1503]: Aug 02 14:47:20.971 INFO Fetching http://168.63.129.16/machine/aa786c20-80a9-4a09-8936-2104c27e6e5c/7f691c1e%2D37bf%2D4b9a%2D8c90%2D010bcc7c4277.%5Fteamcity?comp=certificates&incarnation=2>
Aug 02 14:47:20 teamcity coreos-metadata[1503]: Aug 02 14:47:20.979 INFO Fetch successful
Aug 02 14:47:20 teamcity coreos-metadata[1503]: thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /build/amd64-usr/var/tmp/portage/dev-libs/rustlib-1.26.0/work/rustc-1.26.0-src/src/libcore/option.rs:335:21
Aug 02 14:47:21 teamcity systemd[1]: coreos-metadata.service: Main process exited, code=exited, status=101/n/a
Aug 02 14:47:21 teamcity systemd[1]: coreos-metadata.service: Failed with result 'exit-code'.
Aug 02 14:47:21 teamcity systemd[1]: Failed to start CoreOS Metadata Agent.
```

```
> cat /usr/lib/systemd/system/coreos-metadata.service
[Unit]
Description=CoreOS Metadata Agent

[Service]
Type=oneshot
Environment=COREOS_METADATA_OPT_PROVIDER=--cmdline
ExecStart=/usr/bin/coreos-metadata ${COREOS_METADATA_OPT_PROVIDER} --attributes=/run/metadata/coreos

[Install]
RequiredBy=metadata.target
```

**Comment by dtretyakov** [comment]
It could be caused by the recent changes made by @csssuf and @sdemos in the: https://github.com/coreos/coreos-metadata/blob/master/src/providers/azure/mod.rs

**Comment by csssuf** [comment]
Those changes haven't made it into a release yet, so this is an existing bug. There are a number of `unwrap()`s in the Azure crypto logic which should probably be removed in favor of proper error reporting, one of which is likely the cause of this issue.

@dtretyakov, can you get a backtrace from `coreos-metadata`? You should be able to do so by putting the following in `/etc/systemd/system/coreos-metadata.service.d/10-backtrace.conf`:
```
[Service]
Environment=RUST_BACKTRACE=1
```

**Comment by sdemos** [comment]
The Azure crypto logic is definitely messy. Doing a `chain_err` instead of `unwrap` is definitely the way to go, I think I left those there initially because of assumptions (that might be wrong) about the cryptographic structures I was getting back from the API. Also, since it's `Option::unwrap` and not `Result::unwrap`, only 3 of the 5 unwrap calls apply (these: https://github.com/coreos/coreos-metadata/blob/master/src/providers/azure/crypto/mod.rs#L66, https://github.com/coreos/coreos-metadata/blob/master/src/providers/azure/crypto/mod.rs#L74, https://github.com/coreos/coreos-metadata/blob/master/src/providers/azure/crypto/mod.rs#L75)

@dtretyakov do you have this machine configured with a public ssh key? If so, what is the key type? 

Also, as a funny side node, this is another situation that will be fixed by https://github.com/coreos/bugs/issues/2362, since all the `unwrap` calls are related to retrieving the public ssh keys but it's causing the `coreos-metadata --attributes` to fail.

**Comment by dtretyakov** [comment]
@sdemos, this machine was created with username/password pair:
![image](https://user-images.githubusercontent.com/1025927/43606794-a80254de-969c-11e8-9f47-422592947c97.png)

So it looks like it may not have expected ssh key.

**Comment by sdemos** [comment]
@dtretyakov if you have the opportunity at some point, I would be very interested in seeing what happens when you get the backtrace as @csssuf suggested. I have a guess as to which unwrap it is triggering (https://github.com/coreos/coreos-metadata/blob/master/src/providers/azure/crypto/mod.rs#L66) but it would be nice to know for sure.

As far as fixing this bug, we should audit all locations where `unwrap` is used, since it's likely not what we actually want to do. However, the [next release of `coreos-metadata`](https://github.com/coreos/coreos-metadata/pull/93) should stop triggering it in your particular case. 

I suspect you can work around this by providing an ssh public key in the web UI. If you still need the username and password to be created, it's possible to do this using ignition in the `passwd` section (see the `passwd` section of the [ignition specification](https://github.com/coreos/ignition/blob/master/doc/configuration-v2_2.md))

**Comment by dtretyakov** [comment]
I've verified that the case with ssh public key works fine.

When was used username/password pair I've tried to enable backtrace for rust but it looks like it does not take any effect:

```
> export RUST_BACKTRACE=1
> /usr/bin/coreos-metadata --cmdline --attributes=/run/metadata/coreos
Aug 03 07:57:35.480 INFO Fetching http://168.63.129.16/?comp=versions: Attempt #1
Aug 03 07:57:35.482 INFO Fetch successful
Aug 03 07:57:35.483 INFO Fetching http://168.63.129.16/machine/?comp=goalstate: Attempt #1
Aug 03 07:57:35.485 INFO Fetch successful
Aug 03 07:57:35.673 INFO Fetching http://168.63.129.16/machine/aadeb670-1ea7-40f2-bee2-a27287cbfc48/86dcad60%2Dca5b%2D45aa%2D91e7%2Df8780e4a7847.%5Fteamcity?comp=certificates&incarnation=2: Attempt #1
Aug 03 07:57:35.682 INFO Fetch successful
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /build/amd64-usr/var/tmp/portage/dev-libs/rustlib-1.26.0/work/rustc-1.26.0-src/src/libcore/option.rs:335:21
> echo $RUST_BACKTRACE
1
```

**Comment by lucab** [comment]
This should be fixed in 2.0.0. For ContainerLinux, it is now released as part of [1871.0.0](https://coreos.com/releases/#1871.0.0).

I'm going ahead and closing this. @dtretyakov feel free to followup if you are still hitting this in the latest release.

**Comment by dtretyakov** [comment]
@lucab, thanks, in CoreOS v1871 it works. Waiting when it will be available in the Stable channel.
